packages/tasit-action/src/TasitAct.js

```diff
@@ -0,0 +1,144 @@
+import "ethers/dist/shims.js";
+// Note: ethers SHOULD be imported from their main object
+// shims aren't injected with package import
+import { ethers } from "ethers";
+
+class Utils {
+  static isAddress = address => {
+    return typeof address === "string" && address.match(/^0x[0-9A-Fa-f]{40}$/);
+  };
+
+  static isABI = abi => {
+    return abi && Array.isArray(abi);
+  };
+
+  // https://github.com/ethers-io/ethers.js/blob/db383a3121bb8cf5c80c5488e853101d8c1df353/src.ts/utils/properties.ts#L20
+  static isEthersJsSigner = signer => {
+    return signer && signer._ethersType === "Signer";
+  };
+}
+
+class Subscription {
+  #txPromise;
+  #provider;
+
+  constructor(txPromise, provider) {
+    this.#txPromise = txPromise;
+    this.#provider = provider;
+  }
+
+  on = async (trigger, callback) => {
+    let blockHeight = 0;
+
+    if (trigger === "confirmation") blockHeight = 1;
+    else {
+      throw new Error(`Invalid subscription trigger, use: ["confirmation"]`);
+    }
+
+    if (!callback || typeof callback !== "function") {
+      throw new Error(`Cannot subscribe without a function`);
+    }
+
+    const tx = await this.#txPromise;
+
+    this.#provider.on(tx.hash, async receipt => {
+      const message = {
+        data: {
+          confirmations: receipt.confirmations,
+        },
+      };
+
+      try {
+        await callback(message);
+      } catch (error) {
+        throw new Error(`Callback function with error: ${error.message}`);
+      }
+    });
+  };
+
+  removeAllListeners = async () => {
+    const tx = await this.#txPromise;
+    this.#provider.removeAllListeners(tx.hash);
+  };
+}
+
+export class Contract {
+  address;
+  #provider;
+  #contract;
+
+  constructor(address, abi, wallet) {
+    if (!Utils.isAddress(address) || !Utils.isABI(abi))
+      throw new Error(`Cannot create a Contract without a address and ABI`);
+
+    this.#provider = this.#getDefaultProvider();
+
+    const signerOrProvider = wallet
+      ? wallet.connect(this.#provider)
+      : this.#provider;
+
+    this.#contract = new ethers.Contract(address, abi, signerOrProvider);
+    this.address = this.#contract.address;
+    this.#addFunctionsToContract();
+  }
+
+  // Note: For now, `tasit-account` creates a wallet object
+  setWallet = wallet => {
+    if (!Utils.isEthersJsSigner(wallet))
+      throw new Error(`Cannot set a invalid wallet to a Contract`);
+
+    return new Contract(this.address, this.#contract.interface.abi, wallet);
+  };
+
+  // For tests purpose
+  getProvider = () => {
+    return this.#provider;
+  };
+
+  // Notes:
+  // - JsonRpcProvider will only be used for testing purpose;
+  // - Default provider should be customized (.env file).
+  #getDefaultProvider = () => {
```

Comment:
I wonder if it's too confusing to call that the default provider if we let them customize it, given that that's the name ethers.js uses specifically for Infura and Etherscan

Reply:
How are you think to customize that? Programatcly, with config/env file or both? We should use a default provider if the user doesn't set up one?

Reply:
Maybe `#getDefaultProvider` gets a hardcoded default, and `#getProviderFromEnv` gets from env, and `#createProvider` takes params programmatically? We may not keep them all, but it's probably simple enough to have options for now. Or we can hold off on some of them. All I feel strongly about on this topic is explicit, clear naming.

Reply:
Related to: https://github.com/tasitlabs/TasitSDK/pull/71#discussion_r243329206

Reply:
I was thinking on that and I've draft a `.env` sample file, what do you think?

```
# Notes:
# - Assuming that IPC won't be used
# - Web3 provider should be receives an object and can be done programmatically only
# - Networks: homestead,rinkeby,ropsten,kovan,other
# - Providers: infura,etherscan,jsonrpc
# - Ethers.js has a FallbackProvider that do the same as default provider (uses more than one provider).
#   This kind of setup will be harder to do thru the .env file

# Example default ganache
# Uses localhost:8545 as default (ethers.js default)
NETWORK=other #(or private)
PROVIDER=jsonrpc

# Example some node (geth/partity)
NETWORK=ropsten
PROVIDER=jsonrpc
JSON_RPC_HOST=192.168.10.20
JSON_RPC_PORT=8545
JSON_RPC_USER=user
JSON_RPC_PASS=pass

# Example infura
NETWORK=homestead #(or mainnet)
PROVIDER=infura
INFURA_API_KEY=xyz # optional

# Example etherscan
NETWORK=rinkeby
PROVIDER=etherscan
ETHERSCAN_API_KEY=zyx # optional

# Example default
# Same as ethers.js, default provider uses both infura and etherscan to improve reliability
# both without API_KEY
NETWORK=kovan
PROVIDER=default
```

Reply:
Reference: https://docs.ethers.io/ethers.js/html/api-providers.html#connecting-to-ethereum

Reply:
@marcelomorgado This looks good to me! Also thanks for including the reference - that's always helpful

Reply:
@marcelomorgado One related idea occurred to me: We can also consider using a `.js` or `.json` config file that is git-ignored. That seems to be a trend over pure `.env` these days

Reply:
Extracted to https://github.com/tasitlabs/TasitSDK/issues/103

Reply:
Great - resolving this